csrc/multidevice/utils.cpp

```diff
@@ -227,6 +227,83 @@ void shardAllLike(TensorView* ref, std::vector<TensorView*> tvs) {
   }
 }
 
+namespace {
+void propagateShardings(
+    const std::vector<TensorView*>& from_tvs,
+    const std::unordered_set<TensorView*>& boundary_tvs,
+    TensorView* ref) {
+  auto tvs_between = scheduler_utils::getAllTvsFrom(from_tvs, boundary_tvs);
+  // std::vector<TensorView*> tvs;
+  // std::copy_if(tvs_between.begin(), tvs_between.end(),
+  // std::back_inserter(tvs), [](TensorView* tv) {return !isDIDReduction(tv);});
+  shardAllLike(ref, {tvs_between.begin(), tvs_between.end()});
+
+  // Remove DID parallelizations on reduction axes.
+  for (auto tv : tvs_between) {
+    for (auto id : tv->getLogicalDomain()) {
+      if (id->isReduction() && id->isDeviceDim()) {
+        id->parallelize(ParallelType::Serial);
+      }
+    }
+  }
+}
+} // namespace
+
+void shardBetween(
+    const std::vector<Expr*>& from,
+    const std::vector<Expr*>& to,
+    TensorView* ref) {
+  std::vector<TensorView*> from_tvs;
```

Comment:
I think this can be implemented as
```
from_tvs = output tensorviews of from;
to_tvs = output tensorviews of to;
shardBetween(from_tvs, to_tvs, ref);
```
This way, it's simpler and leaves the slightly error-prone logic of grabbing the input tensorviews of `from` only in the other helper.

Reply:
`getAllTvsFrom` adds all reachable TVs, so we have to add the inputs of `from` exprs into the boundary.

Reply:
I understood that, and that's why I suggested call the TV version of `shardBetween` and not `propagateShardings` directly. Am I missing something?

Reply:
Sorry, I got confused with the different names and thought you proposed removing the inputs of `from`, you're suggestion should work